cloudinit/config/cc_resizefs.py

```diff
@@ -88,56 +85,23 @@ def _resize_zfs(mount_point, devpth):
     return ('zpool', 'online', '-e', mount_point, devpth)
 
 
-def _get_dumpfs_output(mount_point):
-    return subp.subp(['dumpfs', '-m', mount_point])[0]
-
-
-def _get_gpart_output(part):
-    return subp.subp(['gpart', 'show', part])[0]
-
-
 def _can_skip_resize_ufs(mount_point, devpth):
-    # extract the current fs sector size
-    """
-    # dumpfs -m /
-    # newfs command for / (/dev/label/rootfs)
-      newfs -L rootf -O 2 -U -a 4 -b 32768 -d 32768 -e 4096 -f 4096 -g 16384
-            -h 64 -i 8192 -j -k 6408 -m 8 -o time -s 58719232 /dev/label/rootf
-    """
-    cur_fs_sz = None
-    frag_sz = None
-    dumpfs_res = _get_dumpfs_output(mount_point)
-    for line in dumpfs_res.splitlines():
-        if not line.startswith('#'):
-            newfs_cmd = shlex.split(line)
-            opt_value = 'O:Ua:s:b:d:e:f:g:h:i:jk:m:o:L:'
-            optlist, _args = getopt.getopt(newfs_cmd[1:], opt_value)
-            for o, a in optlist:
-                if o == "-s":
-                    cur_fs_sz = int(a)
-                if o == "-f":
-                    frag_sz = int(a)
-    # check the current partition size
-    # Example output from `gpart show /dev/da0`:
-    # =>      40  62914480  da0  GPT  (30G)
-    #         40      1024    1  freebsd-boot  (512K)
-    #       1064  58719232    2  freebsd-ufs  (28G)
-    #   58720296   3145728    3  freebsd-swap  (1.5G)
-    #   61866024   1048496       - free -  (512M)
-    expect_sz = None
-    m = re.search('^(/dev/.+)p([0-9])$', devpth)
-    gpart_res = _get_gpart_output(m.group(1))
-    for line in gpart_res.splitlines():
-        if re.search(r"freebsd-ufs", line):
-            fields = line.split()
-            expect_sz = int(fields[1])
-    # Normalize the gpart sector size,
-    # because the size is not exactly the same as fs size.
-    normal_expect_sz = (expect_sz - expect_sz % (frag_sz / 512))
-    if normal_expect_sz == cur_fs_sz:
-        return True
-    else:
-        return False
+    # possible errors cases on the code-path to growfs -N following:
+    # https://github.com/freebsd/freebsd/blob/HEAD/sbin/growfs/growfs.c
+    # This is the "good" error:
+    skip_start = "growfs: requested size"
+    skip_contain = "is not larger than the current filesystem size"
+    # growfs exits with 1 for almost all cases up to this one.
+    # This means we can't just use rcs=[0, 1] as subp parameter:
```

Comment:
I think this implementation is cleaner than the `rcs` one, but I'm not sure I understand this comment. What was preventing the rcs version from working?

Reply:
look at [growfs.c](https://github.com/freebsd/freebsd/blob/master/sbin/growfs/growfs.c#L1373):
every single [err/errc/errx](https://www.freebsd.org/cgi/man.cgi?err(3)) in main () is called with 1

so while it was easy to enumerate the one good case, we would have either had to log (the function contact currently has no log object) or throw an error, because we cannot guarantee that `growfs -y` won't break the FS

an alternative would've been to lie: claiming that we can skip the resize if something is broken, and i really didn't have that in my heart

Reply:
a patch to FreeBSD to add different exit values https://reviews.freebsd.org/D27161